e2e-testing/readme.md

```diff
@@ -0,0 +1,164 @@
+
+![E2E Testing Pyramid](./images/e2e-pyramid.png)
+
+The above illustration is a testing pyramid from [Kent C. Dodd’s blog](https://blog.kentcdodds.com/write-tests-not-too-many-mostly-integration-5e8c7fff591c) which is a combination of the pyramids from [Martin Fowler’s blog](https://martinfowler.com/bliki/TestPyramid.html) and the [Google Testing Blog](https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html).
+
+The majority of your tests are at the bottom of the pyramid. As you move up the pyramid, the number of tests gets smaller. Also, going up the pyramid, tests get slower and more expensive to write, run, and maintain. Therefore, before delving yourself in E2E testing specifications and planning, it is important to keep the checklist of the ***essential activities that should be included in every E2E process***:
+
+- A thorough study of requirements to perform E2E testing.
+- Proper set up of test environments.
+- A thorough study of Hardware and Software requirements.
+- Descriptions of all subsystems as well as main software system involved.
+- Enlist the roles and responsibilities for all the systems and subsystems involved.
+- Testing methods used under E2E testing as well as standards that are followed, its description.
+- Test cases designing as well as tracing requirement matrix.
+- Record or save the input and output data for each system.
+
+When working on large and complex customer engagements, it is very important to perform E2E testing in a controlled environment, and for a good reason. Even  slight change in the sub-system implementation, the interface contract, data schema, etc. can cause interface conflicts. Therefore, we need **constantly** and **automatically** control E2E tests to manage the scope and breadth of the conflicts.
+
+## E2E Testing Methods
+
+### Horizontal Test
+
+This method is used very commonly. It occurs horizontally across the context of multiple applications. Take an example of a web-based application of an online ordering system. The whole process will include accounts, inventory status of the products as well as shipping details.
+
+### Vertical Test
+
+In this method, all the transactions of any application are verified and evaluated right from the start to finish. Each individual layer of the application is tested starting from top to bottom. Take an example of a web-based application that uses middleware API for reaching back-end services. In such case, API is required to be fully tested against both, the underlying back-end services and the top layer UI web interface. All these complex testing scenarios will require proper validation and dedicated automated testing. Thus this method is much more difficult.
```

Comment:
"all the transactions"... it may be interested to talk about 'the most valuable/risky transactions'... and this leads to the overall topic of how to get started with e2e for existing systems, eg. start with one, deep e2e test, and then successively add more-- in some sensible order.

Reply:
good point. I've replaced it with "all most critical transactions"